### PR TITLE
Added test for expired manifest [Repository Rewrite]

### PR DESCRIPTION
**Pull request by ColeHiggins2**
added tests for expired manifest.

Refreshing an expired manifest should fail

### CODE REVIEW

**Comment by ColeHiggins2** [comment]
Syncing with the Candlepin team and Support to find other "expired manifest" use cases

File: tests/foreman/api/test_repositories.py [hunk 1]
```diff
@@ -138,3 +142,28 @@ def test_positive_epel_repositories_with_mirroring_policy(
     )[0]
     assert repodata.content_counts['rpm'] != 0
     assert module_repo_options['mirroring_policy'] == repodata.mirroring_policy
+
+
+def test_negative_uplaod_expired_manifest(module_org, target_sat):
```
Comment (adarshdubey-star) [inline tests/foreman/api/test_repositories.py:147]:
```suggestion
def test_negative_upload_expired_manifest(module_org, target_sat):
```

File: tests/foreman/api/test_repositories.py [hunk 2]
```diff
@@ -138,3 +142,28 @@ def test_positive_epel_repositories_with_mirroring_policy(
...
+        1. Upload a manifest
+        2. Delete the Subscription Allocation on RHSM
+        3. Attempt to refresh the manifest
+
+    :expectedresults: Manifest refresh should fail and return error message
+    """
+    manifester = Manifester(manifest_category=settings.manifest.entitlement)
+    manifest = manifester.get_manifest()
+    manifester.delete_subscription_allocation()
+    upload_manifest(module_org.id, manifest.content)
```
Comment (adarshdubey-star) [inline tests/foreman/api/test_repositories.py:162]:
```suggestion
    target_sat.upload_manifest(module_org.id, manifest.content)
```
Lets use the target_sat if its present : )

**Comment by synkd** [comment]
trigger: test-robottelo
pytest: tests/foreman/api/test_repositories.py::test_negative_uplaod_expired_manifest

**Review by synkd — CHANGES_REQUESTED** [review]
Looks like @adarshdubey-star is correct about `upload_manifest`. It's no longer located in `api/utils`, so PRT is failing on test collection. Implementing Adarsh's suggestions should correct this error.

**Review by pondrejk — APPROVED** [review]

File: tests/foreman/api/test_repositories.py [hunk 3]
```diff
@@ -138,3 +141,28 @@ def test_positive_epel_repositories_with_mirroring_policy(
...
+        1. Upload a manifest
+        2. Delete the Subscription Allocation on RHSM
+        3. Attempt to refresh the manifest
+
+    :expectedresults: Manifest refresh should fail and return error message
+    """
+    manifester = Manifester(manifest_category=settings.manifest.entitlement)
+    manifest = manifester.get_manifest()
+    manifester.delete_subscription_allocation()
+    target_sat.upload_manifest(module_org.id, manifest.content)
```
Comment (Griffin-Sullivan) [inline tests/foreman/api/test_repositories.py:161]: In the docstring it says upload first then delete the allocation, but in the test we are doing the opposite. Should we update the docstring to reflect the test's steps?

**Review by Griffin-Sullivan — APPROVED** [review]
One minor comment. Let's rebase and have someone merge this once the checks pass.

**Review by adarshdubey-star — APPROVED** [review]
Ack , please check the CI failures

**Comment by JacobCallahan** [comment]
@ColeHiggins2 ping on this

**Comment by ColeHiggins2** [comment]
Working on updated code and addressing comments

File: tests/foreman/api/test_repositories.py (hunk 3 above)
Comment (ColeHiggins2) [reply tests/foreman/api/test_repositories.py]: technically this work is done in the backend though manifester so it doesn't matter, but I will update it in code so it reads better

**Comment by ColeHiggins2** [comment]
trigger: test-robottelo
pytest: tests/foreman/api/test_repositories.py -k test_negative_upload_expired_manifest

File: tests/foreman/api/test_repositories.py [hunk 4]
```diff
@@ -17,11 +17,14 @@
...
 """
 import pytest
+from manifester import Manifester
 from nailgun import entities
 from nailgun.entity_mixins import call_entity_method_with_timeout
 from requests.exceptions import HTTPError
 
 from robottelo import constants
+from robottelo.cli.base import CLIReturnCodeError
+from robottelo.cli.subscription import Subscription
```
Comment (jyejare) [inline tests/foreman/api/test_repositories.py:27]: These are need not to be imported now simply do :

```suggestion
target_sat.cli.Subscription
```

File: tests/foreman/api/test_repositories.py [hunk 5]
```diff
@@ -190,3 +193,28 @@ def test_positive_sync_kickstart_repo(self, module_entitlement_manifest_org, tar
     rh_repo = rh_repo.read()
     assert rh_repo.content_counts['package_group'] > 0
     assert rh_repo.content_counts['rpm'] > 0
+
+
+@pytest.mark.tier1
+def test_negative_upload_expired_manifest(module_org, target_sat):
+    """Upload an expired manifest and attempt to refresh it
```
Comment (jyejare) [inline tests/foreman/api/test_repositories.py:200]: Do you think the case level and case importance is right here thats being fetched from module hierarchy?

File: tests/foreman/api/test_repositories.py [hunk 6]
```diff
@@ -190,3 +193,28 @@ def test_positive_sync_kickstart_repo(self, module_entitlement_manifest_org, tar
...
+
+    :id: d6e652d8-5f46-4d15-9191-d842466d45d0
+
+    :Steps:
+        1. Upload a manifest
```
Comment (jyejare) [inline tests/foreman/api/test_repositories.py:205]:
```suggestion
        1. Upload an expired manifest
```

**Review by jyejare — APPROVED** [review]
Some minor comments and nitpicks to address.

**Review by vsedmik — APPROVED** [review]

File: tests/foreman/api/test_repositories.py (hunk 6 above)
Comment (ColeHiggins2) [reply tests/foreman/api/test_repositories.py:205]: we are not uploading an expired manifest here, we upload a manifest, then delete the subscription allocation causing the manifest to be expired.

File: tests/foreman/api/test_repositories.py (hunk 5 above)
Comment (ColeHiggins2) [reply tests/foreman/api/test_repositories.py:200]: most tests prior have been marked to tier1 due to these being support cases. If you think I should remove this, should I remove for all other tests too ?

**Review by synkd — APPROVED** [review]

**Comment by ogajduse** [comment]
trigger: test-robottelo
pytest: tests/foreman/api/test_repositories.py -k test_negative_upload_expired_manifest

**Review by ogajduse — APPROVED** [review]
Giving 7th ACK and merging as PRT passed.

**Pull request merged**